runtime/js/99_main.js

```diff
@@ -89,6 +95,93 @@ ObjectDefineProperties(Symbol, {
 let windowIsClosing = false;
 let globalThis_;
 
+let deprecatedApiWarningDisabled = false;
+const ALREADY_WARNED_DEPRECATED = new SafeSet();
+
+function warnOnDeprecatedApi(apiName, stack, suggestion) {
+  if (deprecatedApiWarningDisabled) {
+    return;
+  }
+
+  if (ALREADY_WARNED_DEPRECATED.has(apiName + stack)) {
+    return;
+  }
+
+  // If we haven't warned yet, let's do some processing of the stack trace
+  // to make it more useful.
+  const stackLines = StringPrototypeSplit(stack, "\n");
+  ArrayPrototypeShift(stackLines);
+  while (true) {
+    // Filter out internal frames at the top of the stack - they are not useful
+    // to the user.
+    if (
+      StringPrototypeIncludes(stackLines[0], "(ext:") ||
+      StringPrototypeIncludes(stackLines[0], "(node:")
+    ) {
+      ArrayPrototypeShift(stackLines);
+    } else {
+      break;
+    }
+  }
+  // Now remove the last frame if it's coming from "ext:core" - this is most likely
+  // event loop tick or promise handler calling a user function - again not
+  // useful to the user.
+  if (
+    StringPrototypeIncludes(stackLines[stackLines.length - 1], "(ext:core/")
+  ) {
+    ArrayPrototypePop(stackLines);
+  }
+
+  let isFromRemoteDependency = false;
+  const firstStackLine = stackLines[0];
+  if (firstStackLine && !StringPrototypeIncludes(firstStackLine, "file:")) {
+    isFromRemoteDependency = true;
+  }
+
+  ALREADY_WARNED_DEPRECATED.add(apiName + stack);
+  console.error(
+    "%cWarning",
+    "color: yellow; font-weight: bold;",
+  );
+  console.error(
+    `%c\u251c Use of deprecated "${apiName}" API.`,
+    "color: yellow;",
+  );
+  console.error("%c\u2502", "color: yellow;");
+  console.error(
+    "%c\u251c This API will be removed in Deno 2.0. Make sure to upgrade to a stable API before then.",
+    "color: yellow;",
+  );
+  console.error("%c\u2502", "color: yellow;");
+  console.error(
+    `%c\u251c Suggestion: ${suggestion}`,
+    "color: yellow;",
+  );
+  if (isFromRemoteDependency) {
+    console.error("%c\u2502", "color: yellow;");
+    console.error(
+      `%c\u251c Suggestion: It appears this API is used by a remote dependency.`,
+      "color: yellow;",
+    );
+    console.error(
+      "%c\u2502             Try upgrading to the latest version of that dependency.",
+      "color: yellow;",
+    );
+  }
+
```

Comment:
We should print the ability to disable these warnings with the `DENO_NO_DEPRECATION_WARNINGS` environment variable.

Reply:
https://github.com/denoland/deno/pull/21939#discussion_r1462640212